src/Analyzers/CSharp/Analyzers/UseUTF8StringLiteral/UseUTF8StringLiteralDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,150 @@
+
+            // If we couldn't get constant values for all elements then we can't offer
+            if (values.Length != arrayCreationExpression.Initializer.ElementValues.Length)
+                return;
+
+            var stringValue = GetStringValue(values);
+
+            if (stringValue is null)
+                return;
+
+            var properties = ImmutableDictionary<string, string?>.Empty.Add(StringValuePropertyName, stringValue);
+
+            context.ReportDiagnostic(
+                DiagnosticHelper.Create(Descriptor, arrayCreationExpression.Syntax.GetLocation(), option.Notification.Severity, additionalLocations: null, properties));
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        private static string? GetStringValue(ImmutableArray<object?> values)
+        {
+            try
+            {
+                var byteValues = new byte[values.Length];
+                for (var i = 0; i < values.Length; i++)
+                {
+                    // We shouldn't get nulls, but Convert.ToByte will return (byte)0 if there are any
+                    // so better to crash if the analyzer has a bug, that output a buggy string in users code.
+                    if (values[i] is null)
+                        throw ExceptionUtilities.Unreachable;
+
+                    byteValues[i] = Convert.ToByte(values[i]);
+                }
+
+                var pooledBuilder = PooledStringBuilder.GetInstance();
+                var builder = pooledBuilder.Builder;
+                var ros = new ReadOnlySpan<byte>(byteValues);
+                for (var i = 0; i < ros.Length;)
+                {
+                    // If we can't decode a rune from the array then it can't be represented as a string
+                    if (Rune.DecodeFromUtf8(ros.Slice(i), out var rune, out var bytesConsumed) != OperationStatus.Done)
+                        return null;
+
+                    i += bytesConsumed;
+                    builder.Append(GetStringLiteralRepresentation(rune));
+                }
+
+                return pooledBuilder.ToStringAndFree();
+            }
+            catch
+            {
+                // Ignore any conversion failures and just don't report the diagnostic
+            }
+
+            return null;
+        }
+
+        private static string GetStringLiteralRepresentation(Rune rune)
+            => rune.Value switch
+            {
+                '"' => "\\\"",
+                '\\' => "\\\\",
+                '\0' => "\\0",
+                '\a' => "\\a",
+                '\b' => "\\b",
+                '\f' => "\\f",
+                '\n' => "\\n",
+                '\r' => "\\r",
+                '\t' => "\\t",
+                '\v' => "\\v",
```

Comment:
Is this list taken from somewhere?

Reply:
https://github.com/dotnet/roslyn/blob/main/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/EmbeddedLanguages/VirtualChars/CSharpVirtualCharService.cs#L254